homeassistant/components/xiaomi_aqara/config_flow.py

```diff
@@ -55,42 +62,69 @@ async def async_step_user(self, user_input=None):
+            # allow optional manual setting of host and mac
+            if self.host is None and self.sid is None:
+                self.host = user_input.get(CONF_HOST)
+                mac_address = user_input.get(CONF_MAC)
+
+                # format sid from mac_address
+                if mac_address is not None:
+                    self.sid = mac_address.replace(":", "").lower()
+
+            # if host is already known by zeroconf discovery or manual optional settings
+            if self.host is not None and self.sid is not None:
+                # Connect to Xiaomi Aqara Gateway
+                self.selected_gateway = await self.hass.async_add_executor_job(
+                    XiaomiGateway,
+                    self.host,
+                    self.sid,
+                    None,
+                    DEFAULT_DISCOVERY_RETRY,
+                    self.interface,
+                    MULTICAST_PORT,
+                    None,
+                )
 
-            if not errors:
-                self.gateways = xiaomi.gateways
+                if self.selected_gateway.connection_error:
+                    errors[CONF_HOST] = "invalid_host"
+                elif self.selected_gateway.mac_error:
+                    errors[CONF_MAC] = "invalid_mac"
+                else:
+                    return await self.async_step_settings()
+            else:
+                # Discover Xiaomi Aqara Gateways in the netwerk to get required SIDs.
+                xiaomi = XiaomiGatewayDiscovery(self.hass.add_job, [], self.interface)
+                try:
+                    await self.hass.async_add_executor_job(xiaomi.discover_gateways)
+                except gaierror:
+                    errors[CONF_INTERFACE] = "invalid_interface"
 
-                # if host is already known by zeroconf discovery
-                if self.host is not None:
-                    self.selected_gateway = self.gateways.get(self.host)
-                    if self.selected_gateway is not None:
-                        return await self.async_step_settings()
+                if not errors:
+                    self.gateways = xiaomi.gateways
 
-                    errors["base"] = "not_found_error"
-                else:
                     if len(self.gateways) == 1:
                         self.selected_gateway = list(self.gateways.values())[0]
+                        self.sid = self.selected_gateway.sid
                         return await self.async_step_settings()
                     if len(self.gateways) > 1:
                         return await self.async_step_select()
 
                     errors["base"] = "discovery_error"
 
-        return self.async_show_form(
-            step_id="user", data_schema=GATEWAY_CONFIG, errors=errors
-        )
+        if (self.host is None and self.sid is None) or errors:
+            schema = GATEWAY_CONFIG_HOST
+        else:
+            schema = GATEWAY_CONFIG
```

Comment:
Avoid the use of else if not needed:

```suggestion
        schema = GATEWAY_CONFIG
        if (self.host is None and self.sid is None) or errors:
            schema = GATEWAY_CONFIG_HOST            
```